src/Stratis.Bitcoin/P2P/Peer/NetworkPeerConnection.cs

```diff
@@ -115,21 +115,37 @@ public NetworkPeerConnection(Network network, INetworkPeer peer, TcpClient clien
         /// <summary>
         /// Starts waiting for incoming messages.
         /// </summary>
-        public void StartReceiveMessages()
+        /// <param name="cancellationToken">Optional cancellation token.</param>
+        public void StartReceiveMessages(CancellationToken cancellationToken = default(CancellationToken))
         {
-            this.receiveMessageTask = this.ReceiveMessagesAsync();
+            this.receiveMessageTask = this.ReceiveMessagesAsync(cancellationToken);
         }
 
         /// <summary>
         /// Reads messages from the connection stream.
         /// </summary>
-        private async Task ReceiveMessagesAsync()
+        /// <param name="cancellationToken">Optional cancellation token.</param>
+        private async Task ReceiveMessagesAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
+            CancellationTokenSource cts = null;
+
             try
             {
-                while (!this.CancellationSource.Token.IsCancellationRequested)
+                // Check if valid cancellation was passed in and link it to the existing cancellation token source
+                if (cancellationToken != default(CancellationToken))
+                {
+                    cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, this.CancellationSource.Token);
+                    cancellationToken = cts.Token;
+                }
+                else
                 {
-                    Message message = await this.ReadAndParseMessageAsync(this.peer.Version, this.CancellationSource.Token).ConfigureAwait(false);
+                    cancellationToken = this.CancellationSource.Token;
```

Comment:
The connector already has an internal cancellation token `this.CancellationSource` I don't think we need to pass it in the method, we just need to allow to access this cancellation from outside.